opt: allow qualified star without FROM clause

RaduBerinde opened the pull request:
We incorrectly error out if we have a qualified star without a FROM clause; the
star can refer to an outer table. We also incorrectly take the columns from the
input scope which is not necessarily the scope that the star is referring to.

Unfortunately, in the specific case in #45855 (`SELECT (SELECT t.*)`) we don't
get the nice column name `x` like in postgres; we get `?column?`. This is
because in our code the aliases are determined before building the expressions;
this seems complicated to change.

Fixes #45855.

Release note (bug fix): queries with qualified stars that refer to tables
in outer scopes now work.

Release justification: low risk changes to existing functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46233)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46233#-:-M2hizaPAR-t1yJ1STUH:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: allow qualified star without FROM clause

We incorrectly error out if we have a qualified star without a FROM clause; the
star can refer to an outer table. We also incorrectly take the columns from the
input scope which is not necessarily the scope that the star is referring to.

Unfortunately, in the specific case in #45855 (`SELECT (SELECT t.*)`) we don't
get the nice column name `x` like in postgres; we get `?column?`. This is
because in our code the aliases are determined before building the expressions;
this seems complicated to change.

Fixes #45855.

Release note (bug fix): queries with qualified stars that refer to tables
in outer scopes now work.

Release justification: low risk changes to existing functionality.

RaduBerinde commented:
bors r+
